src/main/java/com/google/sps/servlets/ImageServlet.java

```diff
@@ -0,0 +1,206 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.protobuf.ByteString;
+import com.google.gson.Gson;
+import java.io.ByteArrayOutputStream;
+import java.net.URL;
+import java.net.MalformedURLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/user-image")
+public class ImageServlet extends HttpServlet {
+  
+  /**
+   * Writes an upload URL for file uploads to the servlet
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstore = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstore.createUploadUrl("/user-image");
+
+    response.getWriter().println(uploadUrl);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PrintWriter writer = response.getWriter();
+    BlobKey blobKey = getBlobKey(request, "image");
+    if (blobKey == null) {
+      writer.println("Please upload an image file.");
+      return;
+    }
+    // TODO: Do something with the URL -- either add it to the writer output to the servlet, 
+    // and make use of it on the website, or remove and ignore.
+    String imageUrl = getUploadedFileUrl(request, "user-image");
```

Comment:
You should probably check that the file is indeed an image, or at least add a TODO or open an Issue.

Reply:
There's a statement earlier in the method that checks whether the file is an image.

Reply:
I think you're checking if there's an image field, but the user could upload a non-image file to the image field.